### PR TITLE
ceph: osd do not use the rook binary for osd on pvc

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

ceph: remove ability to create rook legacy rook

This commits removes all the function/methods/tests that were used to
create and manage rook legacy OSDS.
It also fixes various go-lint issues in the respective files.

ceph: implement c-v raw mode for osd on pvc

Multiple things:

1. use c-v inventory to detect available devices:
Now we rely on the 'ceph-volume inventory' command to tell us if a
device is available or not.

2. allow the use of partitions for osds:
If specified in the cluster CR, a partition will be used to create the
OSD. This is also working when using "useAllDevices: true". The
partition will be preferred over the main disk, if multiple partitions
from the same devices are discovered they will be used too.

3. implement raw mode for osd on pvc
When an OSD will be bootstrap on a PVC, the new c-v raw mode will be
used. It consists of putting block, db and wal under the same device.
Here LVM is out of the picture and the raw device is used as is. The
implementation is backward compatible so existing OSD on PVC will LVM
will continue to operate.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves https://github.com/rook/rook/issues/4363 and https://github.com/rook/rook/issues/4724

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test ceph]
[test full]

### CODE REVIEW

**Comment by BlaineEXE** [comment]
Is this still WIP, based on the description?

**Comment by leseb** [comment]
> Is this still WIP, based on the description?

Oops, you can go ahead :). Thanks.

File: pkg/clusterd/disk.go [hunk 1]
```diff
@@ -50,24 +51,49 @@ func DiscoverDevices(executor exec.Executor) ([]*sys.LocalDisk, error) {
...
 			continue
 		}
+
+		// Populate device information coming from lsblk
 		disk, err := PopulateDeviceInfo(d, executor)
 		if err != nil {
-			// skip device
-			logger.Warningf("skipping device %s: %+v", d, err)
+			// skip device if lsblk fails
+			logger.Warningf("skipping device %q because 'lsblk' failed. %+v", d, err)
```
Comment (BlaineEXE) [inline pkg/clusterd/disk.go]: Do we need to be modifying these so they use `%v` instead of `%+v`?

File: pkg/clusterd/disk.go [hunk 2]
```diff
@@ -50,24 +51,49 @@ func DiscoverDevices(executor exec.Executor) ([]*sys.LocalDisk, error) {
...
+		// and means it's a parent block device
+		if disk.Type == sys.DiskType {
+			deviceChild, err := sys.ListDevicesChild(executor, d)
+			if err != nil {
+				logger.Warningf("failed to detect child devices for device %q. %+v", d, err)
+			}
+			// lsblk will output at least 2 lines if they are partitions, one for the parent
+			// and N for the child
+			if len(deviceChild) > 1 {
+				logger.Warningf("skipping device %q because it has child, considering the child instead.", d)
```
Comment (BlaineEXE) [inline pkg/clusterd/disk.go]: This seems like it would be info to me instead of warning.

File: pkg/daemon/ceph/osd/volume_test.go [hunk 3]
```diff
@@ -238,21 +239,57 @@ var cephVolumeTestResultMultiClusterMultiOSD = `{
...
 `
 
-func TestParseCephVolumeResult(t *testing.T) {
+var cephVolumeRAWTestResult = `{
+    "0": {
+        "ceph_fsid": "4bfe8b72-5e69-4330-b6c0-4d914db8ab89",
+        "device": "/dev/vdb",
+        "osd_id": 0,
+        "osd_uuid": "c03d7353-96e5-4a41-98de-830dfff97d06",
+        "type": "bluestore"
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/volume_test.go]: FWIW, the OSD code no longer obeys `type` and always creates bluestore OSDs. This should be removed from the API spec, but I found this to be one of the most difficult parts of stripping out all the OSD code for legacy OSDs. We might do this in a separate PR.

File: pkg/util/sys/device.go [hunk 4]
```diff
@@ -82,8 +103,11 @@ type LocalDisk struct {
 	Empty bool `json:"empty"`
 	// Information provided by Ceph Volume Inventory
 	CephVolumeData string `json:"cephVolumeData,omitempty"`
+	// RealName is the device behind the PVC, behind /mnt/<pvc/name
```
Comment (BlaineEXE) [inline pkg/util/sys/device.go]:
```suggestion
	// RealName is the device behind the PVC, behind /mnt/<pvc>/name
```

File: pkg/util/sys/device.go [hunk 5]
```diff
@@ -280,25 +279,12 @@ func GetDiskUUID(device string, executor exec.Executor) (string, error) {
...
-	return parsePartLabel(output), nil
-}
-
+// MountDevice mount a device
 func MountDevice(devicePath, mountPath string, executor exec.Executor) error {
 	return MountDeviceWithOptions(devicePath, mountPath, "", "", executor)
 }
 
-// comma-separated list of mount options passed directly to mount command
+// MountDeviceWithOptions comma-separated list of mount options passed directly to mount command
```
Comment (BlaineEXE) [inline pkg/util/sys/device.go]: Add more detail than just the name?
```suggestion
// MountDeviceWithOptions mounts a device with a comma-separated list of mount options passed directly to  'mount' command
```

File: pkg/clusterd/disk.go [hunk 6]
```diff
@@ -41,7 +42,7 @@ func ignoreDevice(d string) bool {
 	return isRBD.MatchString(d)
 }
 
-// Discover all the details of devices available on the local node
+// DiscoverDevices all the details of devices available on the local node
```
Comment (BlaineEXE) [inline pkg/clusterd/disk.go]: More grammatical completion to sentence?

```suggestion
// DiscoverDevices returns all the details of devices available on the local node
```

**Comment by BlaineEXE** [comment]
`remove ability to create legacy rook` (commit subj) doesn't seem to be a full statement.

**Comment by BlaineEXE** [comment]
>     1. use c-v inventory to detect available devices:
>        Now we rely on the 'ceph-volume inventory' command to tell us if a
>        device is available or not.

I gathered from comments this is no longer true?

> 
>     2. allow the use of partitions for osds:
>        If specified in the cluster CR, a partition will be used to create the
>        OSD. This is also working when using "useAllDevices: true". The
>        partition will be preferred over the main disk, if multiple partitions
>        from the same devices are discovered they will be used too.

Isn't this in a separate PR now?

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 7]
```diff
@@ -269,64 +240,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 
 	commonArgs = append(commonArgs, osdOnSDNFlag(c.Network, c.clusterInfo.CephVersion)...)
 
+	// RE-ADD OSD ON DIRECTORY?
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: Nope, agreed to remove them!

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 8]
```diff
@@ -269,64 +240,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
...
-			"--osd-max-object-name-len", "256",
-			"--osd-max-object-namespace-len", "64",
-			"--crush-location", fmt.Sprintf("root=default host=%s", osdProps.crushHostname),
-			// Set '--setuser-match-path' so that existing directory owned by root won't affect the daemon startup.
-			// For existing data store owned by root, the daemon will continue to run as root
-			"--setuser-match-path", osd.DataPath,
-		)
-
-		// If the Bluestore OSD was prepared with ceph-volume and not running on PVC
-	} else if osd.CephVolumeInitiated && osdProps.pvc.ClaimName == "" && !osd.IsFileStore {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: We still need this case for a bluestore OSD created with c-v, right? Or maybe it just simplified/moved somewhere else

File: pkg/operator/ceph/cluster/osd/config/scheme.go [hunk 9]
```diff
@@ -13,504 +13,17 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-type PartitionType int
+package config
 
 const (
-	WalPartitionType PartitionType = iota
-	DatabasePartitionType
-	BlockPartitionType
-	FilestoreDataPartitionType
+	// Filestore represents Bluestore
+	Filestore = "filestore"
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/config/scheme.go]: Not if we get ride of Filestore :)

File: pkg/daemon/ceph/osd/daemon.go [hunk 10]
```diff
@@ -226,49 +230,44 @@ func Provision(context *clusterd.Context, agent *OsdAgent, crushLocation string)
...
+
+		// If empty the osd is using the ceph-volume raw mode
+		// so it's consumming a raw block device and LVM is not used
+		// so there is nothing to de-activate
+		if volumeGroupName != "" {
+			if err := releaseLVMDevice(context, volumeGroupName); err != nil {
+				return errors.Wrap(err, "failed to release device from lvm")
+			}
+		} else {
+			// TODO
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/daemon/ceph/osd/daemon.go]: It'd be nice but can be a follow up too.

File: pkg/daemon/ceph/osd/daemon.go [hunk 11]
```diff
@@ -58,18 +55,21 @@ func StartOSD(context *clusterd.Context, osdType, osdID, osdUUID, lvPath string,
...
+
 		go handleTerminate(context, lvPath, volumeGroupName)
 
-		if err := context.Executor.ExecuteCommand(false, "", "vgchange", "-an", volumeGroupName); err != nil {
-			return errors.Wrapf(err, "failed to deactivate volume group for lv %q", lvPath)
-		}
+		// If empty the osd is using the ceph-volume raw mode
+		// so it's consumming a raw block device and LVM is not used
+		// so there is nothing to de-activate
+		if volumeGroupName != "" {
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/daemon/ceph/osd/daemon.go]: Good catch!

File: pkg/util/sys/device.go (hunk 5 above)
Comment (leseb) [reply pkg/util/sys/device.go]: This function is gone now :)

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 8 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: It's in the else :)

**Comment by leseb** [comment]
> > ```
> > 1. use c-v inventory to detect available devices:
> >    Now we rely on the 'ceph-volume inventory' command to tell us if a
> >    device is available or not.
> > ```
> 
> I gathered from comments this is no longer true?
> 
> > ```
> > 2. allow the use of partitions for osds:
> >    If specified in the cluster CR, a partition will be used to create the
> >    OSD. This is also working when using "useAllDevices: true". The
> >    partition will be preferred over the main disk, if multiple partitions
> >    from the same devices are discovered they will be used too.
> > ```
> 
> Isn't this in a separate PR now?

Fixed thanks!

**Comment by travisn** [comment]
LGTM with the following suggestions before it's merged.
- There is so much changing here around OSDs, we need to clear about all the test scenarios that should be passing (at least manually) before it's merged. My thoughts on scenarios are below.
- I've enabled `[test full]` and rerun the CI so we can get a better coverage. There are a number of random CI failures that I don't think will impact the core of these changes in the integration tests. The upgrade test is the area most likely to fail in CI so hopefully it doesn't mask some issue. 
- The upgrade integration test may actually fail every time with these changes since it now upgrades from mimic legacy OSDs to nautilus, and at the end upgrades to master where we don't support the legacy OSDs anymore with this PR. Did you update that integration test? @BlaineEXE can confirm for sure if this needs updating.
- As discussed, we need an issue opened with the migration plan for legacy OSDs so we can get the word out to start transitioning before 1.3. Input from @galexrt @BlaineEXE on the migration instructions would be very helpful to start. Eventually these instructions would be included in the 1.3 upgrade guide.

Test scenarios:
1. OSD on a raw device
1. OSD on a partition
1. OSD on a PVC
1. OSD on a raw device that was created in v1.2 with c-v on LVM. Upgrade to 1.3 and it would still be running on LVM.
1. OSDs created with useAllDevices, deviceFilter, or individually listed devices under the `nodes` section of the CR. I'm not sure this PR will impact those options much, except perhaps in the osd prepare pod. 
1. OSD on a directory. Will they get an error message in the log or will it just be ignored by the operator now? I'm thinking an error in the log would be helpful so if someone misses all the announcements they will hopefully see an error in the log.

**Comment by leseb** [comment]
This is good to go for me, testing done:

### TEST 1 - OSD RUNNING ON DIRECTORY

After the update, the directory OSD keeps running as expected.
Then edited the CR to enable `useAllDevices: true`
As expected a new OSD came up on the drive.

### TEST 2 - RUN ON PVC - UPGRADE FROM LVM

OSD on PVC with LVM, upgraded the operator to the new image, LVs are still running.

### TEST 3 - RUN ON PVC - NEW CLUSTER

OSDs showed up using the c-v raw mode.

### TEST 4 - PARTITION

This only works when running Ceph Octopus... as describe https://github.com/rook/rook/pull/4682#issuecomment-575180057

### TEST 5 - OSD on a raw device that was created in v1.2 with c-v on LVM. 

Upgrade to 1.3 and it is still running on LVM.
OK.

**Comment by BlaineEXE** [comment]
From reading this, I don't see clearly any indication that legacy filestore/bluestore OSDs (created with Mimic <13.2.3 and Rook < v1.1) are still supported or no longer supported. I know that we can no longer **create** legacy OSDs. Can we still run them?

**Comment by leseb** [comment]
Ok, I managed to get a decently cleaned CI, see:

Screenshot 2020-01-24 at 15 25 37


The test on Kubernetes 1.13 failed with:

```

2020-01-24 14:02:57.102711 I | testutil: Running kubectl [apply -f -]

The CephObjectStore "default" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"ceph.rook.io/v1", "kind":"CephObjectStore", "metadata":map[string]interface {}{"name":"default", "namespace":"helm-ns", "creationTimestamp":"2020-01-24T14:02:57Z", "annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"ceph.rook.io/v1\",\"kind\":\"CephObjectStore\",\"metadata\":{\"annotations\":{},\"name\":\"default\",\"namespace\":\"helm-ns\"},\"spec\":{\"dataPool\":{\"replicated\":{\"size\":1}},\"gateway\":{\"allNodes\":false,\"instances\":3,\"port\":80,\"securePort\":null,\"sslCertificateRef\":null,\"type\":\"s3\"},\"metadataPool\":{\"replicated\":{\"size\":1}}}}\n"}, "generation":1, "uid":"39323601-3eb2-11ea-a834-0e5e80dc7ed1"}, "spec":map[string]interface {}{"dataPool":map[string]interface {}{"replicated":map[string]interface {}{"size":1}}, "gateway":map[string]interface {}{"securePort":interface {}(nil), "sslCertificateRef":interface {}(nil), "type":"s3", "allNodes":false, "instances":3, "port":80}, "metadataPool":map[string]interface {}{"replicated":map[string]interface {}{"size":1}}}}: validation failure list:

spec.gateway.securePort in body must be of type integer: "null"

2020-01-24 14:02:57.297502 I | Exit Status: 1

2020-01-24 14:02:57.297547 E | utils: Failed to execute stdin: kubectl [apply -f -] : exit status 1

2020-01-24 14:02:57.297563 E | testutil: Failed to execute kubectl [apply -f -] -- Failed to run stdin: kubectl [apply -f -] : The CephObjectStore "default" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"ceph.rook.io/v1", "kind":"CephObjectStore", "metadata":map[string]interface {}{"name":"default", "namespace":"helm-ns", "creationTimestamp":"2020-01-24T14:02:57Z", "annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"ceph.rook.io/v1\",\"kind\":\"CephObjectStore\",\"metadata\":{\"annotations\":{},\"name\":\"default\",\"namespace\":\"helm-ns\"},\"spec\":{\"dataPool\":{\"replicated\":{\"size\":1}},\"gateway\":{\"allNodes\":false,\"instances\":3,\"port\":80,\"securePort\":null,\"sslCertificateRef\":null,\"type\":\"s3\"},\"metadataPool\":{\"replicated\":{\"size\":1}}}}\n"}, "generation":1, "uid":"39323601-3eb2-11ea-a834-0e5e80dc7ed1"}, "spec":map[string]interface {}{"dataPool":map[string]interface {}{"replicated":map[string]interface {}{"size":1}}, "gateway":map[string]interface {}{"securePort":interface {}(nil), "sslCertificateRef":interface {}(nil), "type":"s3", "allNodes":false, "instances":3, "port":80}, "metadataPool":map[string]interface {}{"replicated":map[string]interface {}{"size":1}}}}: validation failure list:spec.gateway.securePort in body must be of type integer: "null"
```

That failure is expected and being fixed in https://github.com/rook/rook/pull/4694.

Moving on and merging.
